source/builtin_macros/src/syntax.rs

```diff
@@ -1339,6 +1534,20 @@ impl VisitMut for Visitor {
             return;
         }
 
+        if let Expr::Binary(ExprBinary { right, .. }) = &expr {
+            if let Expr::Matches(ExprMatches {
+                op_expr: Some(MatchesOpExpr { op_token, .. }),
+                ..
+            }) = &**right
+            {
+                if matches!(op_token, MatchesOpToken::Implies(_)) {
+                    *expr = Expr::Verbatim(
+                        quote_spanned! { expr.span() => compile_error!("matches using ==> is currently not allowed on the right-hand-side of && and || (use parentheses)") },
```

Comment:
This check may need to be generalized.  For example:

```
assert(false == E::A matches E::B && false);
```

Reply:
I generalized the check, and added a number of tests.